### ISSUE TITLE
Unable to run with a only a rspec "include" tags specified, they are ignored and all non-matching specs are run

### ISSUE THREAD

**Issue by thewoolleyman** [KnapsackPro/knapsack_pro-ruby#37]
Hello,

The README says you can do this, with the following example:

```
# run only tests with tagA
KNAPSACK_PRO_TEST_SUITE_TOKEN_RSPEC=api_key_for_tagA bundle exec rake "knapsack_pro:rspec[--tag tagA]"
```

But we are not able to make it work.   When we specify only two single tags to run, they is ignored with the message **"All examples were filtered out; ignoring..."** and rspec ends up running everything:

```
I, [2017-05-09T00:22:17.628995 #793]  INFO -- : [knapsack_pro] bundle exec rspec -t uses_extra_session --format progress  --default-path spec "spec/selenium/projects/stories/comments_spec.rb"
D, [2017-05-09T00:22:18.152576 #793] DEBUG -- : [knapsack_pro] Test suite time execution queue recording enabled.
Run options: include {:focus=>true, :uses_extra_session=>true}
 
All examples were filtered out; ignoring {:focus=>true, :uses_extra_session=>true}
...
<Knapsack Pro ends up running all non-matching examples>
```

We tried setting the following Rspec config:
`config.run_all_when_everything_filtered = false`

...but this did not help.  Plus, this option is going away anyway (see https://github.com/rspec/rspec-core/issues/2266).

This is with Rspec v 3.4.4.

Thanks,
-- Chad

**Comment by ArturT** [comment]
Hi Chad,

You use knapsack_pro regular mode for that, right?

I tested it and I have RSpec 3.6.0 and it executes only specified tag:

```
bundle exec rake "knapsack_pro:rspec[--tag tagA]"
```

```
# spec/tags_spec.rb
describe 'Tags' do
  it 'a green test', :tagA do
    sleep 2
  end

  it 'b green test', :tagB do
    sleep 4
  end
end
```

```
# rspec output
D, [2017-05-09T09:41:28.842772 #28856] DEBUG -- : [knapsack_pro] API request UUID: c1fcbb7c-86bf-4ff1-b85d-e75d46e390b7
D, [2017-05-09T09:41:28.842865 #28856] DEBUG -- : [knapsack_pro] Test suite split seed: de0bff75-5c35-491d-8469-09c7a72bebd7
D, [2017-05-09T09:41:28.842886 #28856] DEBUG -- : [knapsack_pro] API response:
D, [2017-05-09T09:41:28.843035 #28856] DEBUG -- : [knapsack_pro] {"build_distribution_id"=>"de0bff75-5c35-491d-8469-09c7a72bebd7", "node_index"=>0, "test_files"=>[{"path"=>"spec/controllers/articles_controller_spec.rb", "time_execution"=>0.04465484619140625}, {"path"=>"spec/controllers/welcome_controller_spec.rb", "time_execution"=>0.013782024383544922}, {"path"=>"spec/services/calculator_spec.rb", "time_execution"=>0.0033693313598632812}, {"path"=>"spec/services/meme_spec.rb", "time_execution"=>0.001837015151977539}, {"path"=>"spec/dir with spaces/foobar_spec.rb", "time_execution"=>0.0007009506225585938}, {"path"=>"spec/foo_spec.rb", "time_execution"=>0.000621795654296875}, {"path"=>"spec/features/calculator_spec.rb", "time_execution"=>0.3324143886566162}, {"path"=>"spec/tags_spec.rb", "time_execution"=>nil}]}
/Users/artur/.rvm/rubies/ruby-2.4.1/bin/ruby -I/Users/artur/.rvm/gems/ruby-2.4.1/gems/rspec-core-3.6.0/lib:/Users/artur/.rvm/gems/ruby-2.4.1/gems/rspec-support-3.6.0/lib /Users/artur/.rvm/gems/ruby-2.4.1/gems/rspec-core-3.6.0/exe/rspec spec/controllers/articles_controller_spec.rb spec/controllers/welcome_controller_spec.rb spec/dir\ with\ spaces/foobar_spec.rb spec/features/calculator_spec.rb spec/foo_spec.rb spec/services/calculator_spec.rb spec/services/meme_spec.rb spec/tags_spec.rb --tag tagA --default-path spec
D, [2017-05-09T09:41:29.281400 #28860] DEBUG -- : [knapsack_pro] Test suite time execution recording enabled.
Run options: include {:tagA=>true}
--------------------Run before test suite--------------------
Tags
  a green test
D, [2017-05-09T09:41:33.174996 #28860] DEBUG -- : [knapsack_pro] API request UUID: ce327fb2-7815-4f0e-9b53-be0d380a1d73
D, [2017-05-09T09:41:33.175098 #28860] DEBUG -- : [knapsack_pro] API response:
D, [2017-05-09T09:41:33.175132 #28860] DEBUG -- : [knapsack_pro]
D, [2017-05-09T09:41:33.175161 #28860] DEBUG -- : [knapsack_pro] Saved time execution report on API server.
D, [2017-05-09T09:41:33.175324 #28860] DEBUG -- : [knapsack_pro] Global time execution for tests: 02s

Finished in 2.11 seconds (files took 1.95 seconds to load)
1 example, 0 failures
```

What RSpec version you have?
What is your knapsack_pro command, please share example.
Please share output of RSpec along with knapsack_pro debug logs.
You can send me email with those private details.

BTW I prepared this tag example long time ago for Regular Mode and it might not be the perfect solution if you want to split big test files and you have a lot CI nodes because you would have to add tags to a lot of test files so each CI node could run something. And if you have 2 tags then you need 3 API keys.

Queue Mode should be good enough to autobalance tests across CI nodes unless you have one very large test file with time execution longer than average time execution for CI node. Is that your case?

Could you describe what problem you are trying to solve?

Cheers,
Artur

**Comment by thewoolleyman** [comment]
No, this is using Knapsack Pro queue mode.  I think I may have been confused and this is **not** supported in queue mode.  Is that correct?

**Comment by ArturT** [comment]
1. I've updated readme to make it more clear it's the example for regular mode.
https://github.com/KnapsackPro/knapsack_pro-ruby#b-use-tags-to-mark-set-of-tests-in-particular-test-file

2. I tested it with Queue Mode today and it works as well but there are cons. If you add tags only for a few test files then in Queue Mode there will be all test files from test suite returned from the Knapsack Pro API. Some of them without tags and they will be skipped.There will be some unnecessary requests to Knapsack Pro API to get test files even they don't have tags. 

You still would need to have 3 API keys if you use 2 tags (tagA, tagB).

3. When I run queue mode then I see in output:

```
Run options: include {:tagA=>true}
```

I'm not sure why in your case it's:
```
All examples were filtered out; ignoring {:focus=>true, :uses_extra_session=>true}
```
Maybe because of different RSpec version or configuration? What's your RSpec version?


4. I would focus on the context of the problem you are trying to solve. Maybe there is a better way I could help. Maybe splitting by tags is not a good solution for the problem you are trying to solve.

Do you have a single test file with time execution longer than average time execution of CI node? If yes then it makes sense to split the long running test file to smaller ones or try to do it with tags. 
Is that your case or something else?

**Comment by thewoolleyman** [comment]
Hi Artur,

Thanks for the prompt responses.

TL;DR; We are still having this problem, but it's not a blocker anymore, and it's also not reproducible by you (perhaps our rspec is out of date) so we can close the issue.

Details and answers to your questions:

It looks like we **are** having this problem with the "regular" mode in addition to queue mode.  I'm also not sure why you are not seeing the "All examples were filtered out" like we are.

To answer your question, we were using this to investigate a CI issue with Capybara/Selenium sessions, and we wanted to split our suite into two separate runs - one which contained specs which used "extra sessions", and one which only contained specs that used one of two "standard" sessions.

So, in this case, rspec tags were the easiest and most logical way to split them up (it was simply a one-line change to each spec file we wanted in the "extra sessions" group.

However, we were able to fix our root issue anyway, even though we never got the KnapsackPro tags to work the way we wanted.

So, as I said above, we can close this issue, and if this limitation becomes an issue again, we'll try upgrading rspec and looking into it more, and if it's still reproducible with all the latest gems we'll re-open this.

Thanks!
-- Chad

**Comment by levicole** [comment]
I'm actually having this problem.

We are splitting our feature specs out from our unit tests. We're in the process of migrating from Cucumber to rspec style feature tests, so we have fewer feature tests right now than we do unit tests. When I run knapsack using the following command:

`knapsack_pro queue:rspec "--tag type:feature"`

My build log is filled with this output:

```
Run options: include {:type=>"feature"}

All examples were filtered out
```

I think the solution is still to not run in queue mode, or in my case, possibly to edit the test file pattern (although, when I edit the test file pattern, I get an error saying that the `test_files` parameter is required even though the glob I'm using returns a list of test files).

It might be worth explicitly stating that running a specific tag is best done not using queue mode. I don't think I saw that in the portion of the readme linked above.

Otherwise, I'm loving knapsack, and thanks for all your hard work!

**Comment by ArturT** [comment]
Hi @levicole , thanks for the feedback. Good timing. We are working on a new feature to [split RSpec slow test files across parallel CI nodes by test examples](https://github.com/KnapsackPro/knapsack_pro-ruby#split-test-files-by-test-cases) and similar bug happened for us. 

With @shadre we are trying to figure out how to force RSpec to run test examples for test files that were already loaded into memory.  I guess this should also help with problem here.

This is still work in progress https://github.com/KnapsackPro/knapsack_pro-ruby/pull/106

I submitted issue to rspec-core repo https://github.com/rspec/rspec-core/issues/2721 to find out some tips. There is example code showing why this bug is happening for Queue Mode.


@levicole Quick tip for you would be for now don't use `~tag` but instead explicitly define directories with tests that should be run in Queue Mode:
https://knapsackpro.com/faq/question/how-to-run-tests-from-a-specific-directory-only-system-tests-or-features-specs
https://knapsackpro.com/faq/question/how-can-i-run-tests-from-multiple-directories
https://knapsackpro.com/faq/question/how-to-run-only-rspec-feature-tests-or-non-feature-tests
https://knapsackpro.com/faq/question/how-to-exclude-tests-from-running-them

**Comment by ArturT** [comment]
It's important to define a list of test files with `KNAPSACK_PRO_TEST_FILE_PATTERN` because knapsack_pro has to send a list of test files from the disk to the API so we can use that to split them. API has to know what test files still exist in the project test suite to be able to know what should be split across parallel CI nodes.

Providing `~tag` to rspec does not tell knapsack_pro what test files you have on the disk. 

So if you would use `KNAPSACK_PRO_TEST_FILE_PATTERN` and `~tag sometag` at the same time it would mean that only a subset of test files detected by `KNAPSACK_PRO_TEST_FILE_PATTERN` will be run because test file has to have `sometag` to be run if you provided `~tag sometag`. It may happen that API will return a file that has no `sometag` in it and then knapsack_pro will not run the test file at all because RSpec will ignore such test file. 

That's why I recommend using `KNAPSACK_PRO_TEST_FILE_PATTERN` and `KNAPSACK_PRO_TEST_FILE_EXCLUDE_PATTERN ` to control if you want to run feature or unit tests.

**Comment by levicole** [comment]
Thanks for responding so quickly! I have tried setting the `KNAPSACK_PRO_TEST_FILE_PATTERN` environment variable, but for some reason, the API responds with an error stating that "test_files parameter is required", which is odd. If I pass the same pattern to `Dir.glob` it returns all of my feature tests.

I've been setting it to something like: `spec/features/**/*_spec.rb`

**Comment by ArturT** [comment]
@levicole Can you send RSpec output logs from CI server to support email https://knapsackpro.com/contact so I could review it and check what is send in requests to API. Please send also how knapsack_pro command looks like in your CI yml config and what CI provider do you use. Thanks.

**Comment by ArturT** [comment]
@levicole Have you resolved the issue? The default pattern for spec is `spec/**{,/*/**}/*_spec.rb`.
Can you share your config to support@knapsackpro.com so we can check on our side what's going on when requests are sent to API. I guess there might be some typo or other configuration that you use that can override the default behavior.

**Comment by levicole** [comment]
Hey @ArturT! We ended up going a different route and just lumping all of our rspec tests together and that worked out fine.

I'll see if I can find some of the older builds and send you the output along with the pattern I was using :). Sorry for taking so long to respond.

**Comment by ArturT** [comment]
Update from me, there is official support for auto split slow RSpec test files on parallel CI nodes. It works with Queue Mode and Regular Mode. Thanks to this you don't need RSpec tags to split your tests.

https://knapsackpro.com/faq/question/how-to-split-slow-rspec-test-files-by-test-examples-by-individual-it

**Comment by chendo** [comment]
We're running into a similar issue. In our test suite, we've split out our suite between basic tests, tests that require selenium, and tests that require selenium and additional containers. The goal of this is to minimise the amount of selenium/additional containers that is need to be started for performance reasons.

Our current folder structure under `spec/features` generally has a file or folder for a particular feature we're trying to test, but each test within this can define tags that opt into selenium and/or additional containers as needed. Ideally, we wouldn't have to move away from this particular structure.

I did create a separate API key per suite thinking this would address the issue, however as per the above discussion, this does not help.

Examples being queued but not run doesn't seem to incur that much overhead, however it would be nice to avoid, which would benefit larger test suites.

It should be possible to use `rspec --dry-run --tag foo --tag ~bar --format json` to figure out which tests to send to Knapsack for consideration.

**Comment by ArturT** [comment]
> It should be possible to use rspec --dry-run --tag foo --tag ~bar --format json to figure out which tests to send to Knapsack for consideration.

@chendo You can provide list of test files or test file example paths like `a_spec.rb[1:1]` with [`KNAPSACK_PRO_TEST_FILE_LIST_SOURCE_FILE`](https://docs.knapsackpro.com/ruby/reference/) to run a given set of tests instead of the whole test suite.

If you use the [RSpec split by test examples feature](https://docs.knapsackpro.com/ruby/split-by-test-examples/) and you don't like how it splits test files by examples then you can set a list of slow test files yourself with [`KNAPSACK_PRO_SLOW_TEST_FILE_PATTERN`](https://docs.knapsackpro.com/ruby/reference/#knapsack_pro_slow_test_file_pattern-internal-rspec).

Please note the [RSpec split by test examples feature](https://docs.knapsackpro.com/ruby/split-by-test-examples/) does not work with RSpec tags due to internal RSpec bug that causes to run unintended tests.

<img width="1037" alt="image" src="https://github.com/KnapsackPro/knapsack_pro-ruby/assets/662545/41816a03-b1e3-4f93-bf6e-b1c8e2961c1f">

**Comment by ArturT** [comment]
We added support for RSpec tags in the RSpec split by test examples feature.

Please update to knapsack_pro [8.0.0](https://github.com/KnapsackPro/knapsack_pro-ruby/blob/master/CHANGELOG.md#800). 
[RSpec split by test examples](https://docs.knapsackpro.com/ruby/split-by-test-examples/) is enabled by default.
